packages/server/src/arpa_reporter/services/validate-upload.js

```diff
@@ -342,6 +351,121 @@ async function validateRules ({ upload, records, rules, trns }) {
 
   return errors
 }
+// Subrecipients can use either the uei, or the tin, or both as their identifier.
+// This helper takes those 2 nullable fields and converts it to a reliable format
+// so we can index and search by them.
+function subrecipientIdString(uei, tin) {
+  if (!uei && !tin) {
+    return '';
+  }
+  return JSON.stringify({ uei, tin })
+}
+
+async function validateReferences({ records }) {
+  const errors = []
+  // These 3 types need to search-able by their unique id so we can quickly verify they exist
+  const projects = {}
+  const subrecipients = {}
+  const awardsGT50k = {}
+
+  const awards = []
+  const expendituresGT50k = []
+  for (const record of records) {
+    switch (record.type) {
+      case 'ec1':
+      case 'ec2':
+      case 'ec3':
+      case 'ec4':
+      case 'ec5':
+      case 'ec7':
+        const projectID = record.content.Project_Identification_Number__c
+        if (projectID in projects) {
+          errors.push(betaValidationWarning(
+            `Project ids must be unique, but another row used the id ${projectID}`))
+        }
+        projects[projectID] = record.content;
+        break
+      case 'subrecipient':
+        const subRecipId = subrecipientIdString(
+          record.content.Unique_Entity_Identifier__c,
+          record.content.EIN__c)
+        if (subRecipId && subRecipId in subrecipients) {
+          errors.push(betaValidationWarning(
+            `Subrecipient ids must be unique, but another row used the id ${subRecipId}`))
+        }
+        subrecipients[subRecipId] = record.content
+        break
+      case 'awards50k':
+        const awardNumber = record.content.Award_No__c
+        if (awardNumber && awardNumber in awardsGT50k) {
```

Comment:
It's a great suggestion, and it did cross my mind. I left it out of this PR because the change is actually a little more complicated than expected, because it overlaps with a different issue/bug that has been reported to us recently (https://github.com/usdigitalresponse/usdr-gost/issues/459)

I'm hoping to add the row/col data into these validations at the same time we fix the rest of the row-number errors in that issue.

Reply:
As a middle-ground, I made sure that these error messages included the raw value from the spreadsheet, so at the very least they can ctrl-f search for that value to find the error.

Reply:
Cool, I think that's a good compromise!